server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java

```diff
@@ -149,7 +167,22 @@ public RetentionLeases(final long primaryTerm, final long version, final Collect
     public RetentionLeases(final StreamInput in) throws IOException {
         primaryTerm = in.readVLong();
         version = in.readVLong();
-        leases = Collections.unmodifiableMap(toMap(in.readList(RetentionLease::new)));
+        final int leasesCount = in.readVInt();
+        switch (leasesCount) {
+            case 0 -> leases = Map.of();
+            case 1 -> {
+                final RetentionLease lease = new RetentionLease(in);
+                leases = Map.of(lease.id(), lease);
+            }
+            default -> {
+                final Map<String, RetentionLease> m = Maps.newLinkedHashMapWithExpectedSize(leasesCount);
+                for (int i = 0; i < leasesCount; i++) {
+                    final RetentionLease lease = new RetentionLease(in);
+                    m.put(lease.id(), lease);
+                }
+                leases = Collections.unmodifiableMap(m);
```

Comment:
Consider using `StreamInput::readMapValues`. May be that one could be improved to have a special case for size=1